kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h

```diff
@@ -1327,6 +1377,119 @@ class ResidualBasedBlockBuilderAndSolver
         }
     }
 
+    /**
+     * @brief This method returns the scale norm considering for scaling the diagonal
+     * @param rA The LHS matrix
+     * @return The scale norm
+     */
+    double GetScaleNorm(TSystemMatrixType& rA)
+    {
+        if (mOptions.Is(NO_SCALING) ) {
+            if (mOptions.Is(CONSIDER_PRESCRIBED_DIAGONAL)) {
+                KRATOS_DEBUG_ERROR_IF(mScaleFactor < std::numeric_limits<double>::epsilon()) << "Scale factor of the diagonal cannot be zero or almost zero" << std::endl;
+                return mScaleFactor;
+            } else {
+                if (mOptions.Is(CONSIDER_NORM_DIAGONAL) ) {
+                    return GetDiagonalNorm(rA)/static_cast<double>(rA.size1());
+                } else {
+                    return GetMaxDiagonal(rA);
+//                     return TSparseSpace::TwoNorm(rA)/static_cast<double>(rA.size1());
+                }
+            }
+        } else {
+            return 1.0;
+        }
+    }
+
+    /**
+     * @brief This method returns the diagonal norm considering for scaling the diagonal
+     * @param rA The LHS matrix
+     * @return The diagonal norm
+     */
+    double GetDiagonalNorm(TSystemMatrixType& rA)
+    {
+        double diagonal_norm = 0.0;
+        #pragma omp parallel for reduction(+:diagonal_norm)
+        for(int i = 0; i < static_cast<int>(TSparseSpace::Size1(rA)); ++i) {
+            diagonal_norm += std::pow(rA(i,i), 2);
+        }
+        return std::sqrt(diagonal_norm);
+    }
+
+    /**
+     * @brief This method returns the diagonal max value
+     * @param rA The LHS matrix
+     * @return The diagonal  max value
+     */
+    double GetAveragevalueDiagonal(TSystemMatrixType& rA)
+    {
+        return 0.5 * (GetMaxDiagonal(rA) + GetMinDiagonal(rA));
+    }
+
+    /**
+     * @brief This method returns the diagonal max value
+     * @param rA The LHS matrix
+     * @return The diagonal  max value
+     */
+    double GetMaxDiagonal(TSystemMatrixType& rA)
+    {
+//         // NOTE: Reduction failing in MSVC
+//         double max_diag = 0.0;
+//         #pragma omp parallel for reduction(max:max_diag)
+//         for(int i = 0; i < static_cast<int>(TSparseSpace::Size1(rA)); ++i) {
+//             max_diag = std::max(max_diag, std::abs(rA(i,i)));
+//         }
+//         return max_diag;
+
+        // Creating a buffer for parallel vector fill
+        const int num_threads = OpenMPUtils::GetNumThreads();
+        Vector max_vector(num_threads, 0.0);
+        #pragma omp parallel for
+        for(int i = 0; i < static_cast<int>(TSparseSpace::Size1(rA)); ++i) {
+            const int id = OpenMPUtils::ThisThread();
+            if (std::abs(rA(i,i)) > max_vector[id])

```

Comment:
OPTIMIZATION OPPORTUNITY: 
current loop accesses rA(i,i) twice per i. Each of this imples a search. We cold halve the searches by simply caching A(i,i)

also if we were to consider the CSR structure of A we could avoid the searches at all

Reply:
I would avoid using CSR structure, as I am afraid it depends of the libray used, and now we are actually considering it's a boost compressed matrix

Reply:
actually in this file we are already assuming that we use ublas...so the dependency already exists (and very heavily)

Reply:
Can we live with the current solution?